mls-rs/src/group/mod.rs

```diff
@@ -4729,6 +4789,419 @@ mod tests {
+            .await
+            .unwrap()
+            .0;
+
+        let mut carol = carol_client
+            .join_group(None, &commit.welcome_messages[0])
+            .await
+            .unwrap()
+            .0;
+
+        // Bob proposes self-remove.
+        let bob_self_remove = bob.propose_self_remove(Vec::new()).await.unwrap();
+
+        // Alice receives the self-remove proposal to be committed.
+        // Carol also receives the self-remove proposal. Carol will need this in order
+        // to process the commit including the proposal, because self-remove proposals
+        // are included by reference.
+        alice
+            .process_incoming_message(bob_self_remove.clone())
+            .await
+            .unwrap();
+        carol
+            .process_incoming_message(bob_self_remove)
+            .await
+            .unwrap();
+
+        // Alice commits Bob's self-remove.
+        let commit = alice.commit(Vec::new()).await.unwrap();
+        alice.apply_pending_commit().await.unwrap();
+
+        // Assert that after applying the commit removing Bob, that Bob is no longer in the group.
+        let expected_members = vec![
+            alice.member_at_index(alice.current_member_index()).unwrap(),
+            carol.member_at_index(carol.current_member_index()).unwrap(),
+        ];
+        itertools::assert_equal(alice.roster().members_iter(), expected_members.clone());
+
+        // Assert that Carol can also process the commit and it removes Bob from the group.
+        carol
+            .process_incoming_message(commit.commit_message.clone())
+            .await
+            .unwrap();
+        itertools::assert_equal(carol.roster().members_iter(), expected_members.clone());
+        // Assert that Bob can process the commit.
+        bob.process_incoming_message(commit.commit_message)
+            .await
+            .unwrap();
+    }
+
+    #[cfg(all(
+        feature = "by_ref_proposal",
+        feature = "custom_proposal",
+        feature = "self_remove_proposal"
+    ))]
+    #[maybe_async::test(not(mls_build_async), async(mls_build_async, crate::futures_test))]
+    async fn commit_with_both_remove_and_self_remove_for_same_client_errors() {
+        let (mut alice, mut bob) = self_remove_group_setup().await;
+
+        // Bob proposes self-remove.
+        let bob_self_remove = bob.propose_self_remove(Vec::new()).await.unwrap();
+
+        // Alice receives the self-remove proposal to be committed.
+        alice
+            .process_incoming_message(bob_self_remove.clone())
+            .await
+            .unwrap();
+
+        // Alice also removes Bob with a regular remove proposal.
+        alice.propose_remove(1, Vec::new()).await.unwrap();
+
+        // Alice commits Bob's self-remove and Alice's removal of Bob. This is an error.
+        let commit = alice.commit(Vec::new()).await;
+        assert_matches!(commit, Err(MlsError::MoreThanOneProposalForLeaf(1)));
```

Comment:
To be consistent with handling of other proposals, the commit should succeed but one of the conflicting proposals should be filtered out. This is so that you don't get stuck - right now Alice can't ever commit unless she clears proposals.

Reply:
Good point -- done with your suggested filter change in the proposal filtering